trace.go

```diff
@@ -296,19 +298,18 @@ type TraceErr struct {
 type Fields map[string]interface{}
 
 type RawTrace struct {
-	Err     json.RawMessage `json:"error"`
-	Traces  `json:"traces"`
-	Message string `json:"message"`
+	Err      json.RawMessage `json:"error"`
+	Traces   `json:"traces"`
+	Message  string   `json:"message"`
+	Messages []string `json:"messages"`
 }
 
 // AddUserMessage adds user-friendly message describing the error nature
-func (e *TraceErr) AddUserMessage(formatArg interface{}, rest ...interface{}) {
+func (e *TraceErr) AddUserMessage(formatArg interface{}, rest ...interface{}) *TraceErr {
 	newMessage := fmt.Sprintf(fmt.Sprintf("%v", formatArg), rest...)
-	if len(e.Message) == 0 {
-		e.Message = newMessage
-	} else {
-		e.Message = strings.Join([]string{e.Message, newMessage}, ", ")
-	}
+	// Prepend so messages higher up the call stack go first.
+	e.Messages = append([]string{newMessage}, e.Messages...)
```

Comment:
nit: this will create a new array. Instead - you could continue to append but reverse the order when necessary.